lupa/_lupa.pyx

```diff
@@ -1821,28 +1821,32 @@ cdef tuple unpack_multiple_lua_results(LuaRuntime runtime, lua_State *L, int nar
 
 cdef void* _lua_alloc_restricted(void* ud, void* ptr, size_t old_size, size_t new_size) nogil:
     # adapted from https://stackoverflow.com/a/9672205
-    cdef size_t* memory_left = <size_t*>ud
+    cdef size_t* memory_left_status = <size_t*>ud
+    cdef size_t memory_left = memory_left_status[0]
+    memory_is_counted = memory_left > 0
 
     if ptr is NULL:
         # <http://www.lua.org/manual/5.2/manual.html#lua_Alloc>:
         # When ptr is NULL, old_size encodes the kind of object that Lua is allocating.
         # Since we don’t care about that, just mark it as 0.
         old_size = 0
 
+    cdef void* new_ptr
     if new_size == 0:
         free(ptr)
-        if memory_left[0] > 0:
-            memory_left[0] += old_size  # add old size to available memory
-        return NULL
+        if memory_is_counted:
+            memory_left += old_size  # add deallocated old size to available memory
+        new_ptr = NULL
```

Comment:
This is never commited to `memory_left_status[0]` because `new_ptr` Is NULL and the assignment at the end of the function is skipped.
```suggestion
            memory_left_status[0] += old_size  # add deallocated old size to available memory
```

Reply:
Hmm. Right. I was intending the final update to depend only on `memory_is_counted`, but that doesn't work in the case that `realloc()` fails. :-/